### PR TITLE
Implement the ability to disable classes

### PR DESCRIPTION
**Pull request by reduz**
* This PR adds the ability to disable classes when building.
* For now it's only possible to do this via command like:
  `scons disable_classes=RayCast2D,Area3D`
* Eventually, a proper UI will be implemented to create a build config file to do this at large scale, as well as detect what is used in the project. Automatic detection of classes used to generate the right build config is also a possibility.
* If needed eventually, the same approach could be done for disabling method registration to remove unused methods.

This PR is part of a series of future ones aimed at customizing release binaries to make them much smaller based on the game.

### CODE REVIEW

**Comment by Calinou** [comment]
> Eventually, a proper UI will be implemented to create a build config file to do this at large scale, as well as detect what is used in the project.

A `custom.py` file can be placed at the Godot source root folder to pass SCons options: https://godot-build-options-generator.github.io/

**Review by touilleMan — APPROVED** [review]
Seems good to me overall, but just a question for my understanding:

This PR allows to strips off class definition from ClassDB.
However it doesn't explicitly remove class definition from the codebase (unlike e.g. using `#ifdef ClassDB_Disable_Foo` macros in the foo.h/foo.cpp files)

So does the smaller binary size comes only from the smaller ClassDB, or does further compiler optimization is expected to entirely strip off the unused classes ?
And how can we be sure the excluded classes' code are stripped off the final binary ? (for instance GDNative class is used when loading the `godot.project` file, so it will be part of the binary even if excluded from ClassDB)

**Comment by Xrayez** [comment]
@reduz did you get inspired by [Goost's existing mechanism of disabling classes](https://goost.readthedocs.io/en/latest/usage/configuring.html#components) that I've implemented myself? 😃

See also:
- https://github.com/goostengine/goost/discussions/49
- https://github.com/goostengine/goost/pull/54
- https://github.com/goostengine/goost/pull/55
- https://github.com/goostengine/goost/pull/86
- https://github.com/goostengine/goost/pull/89

**Comment by neikeq** [comment]
> This PR is part of a series of future ones aimed at customizing release binaries to make them much smaller based on the game.

I don't think this is the right approach considering that's the goal. I'll explain more in deep later today. Please wait before merging.

**Comment by reduz** [comment]
@touilleMan if the class is not used by the binder, this means the class does not exist (unless the engine uses it for something internally). The linker builds a dependency tree between symbols and anything not connected to the entry point is removed.

@Xrayez Sorry, I don't pay attention to Goost, I have no idea what is going on in there.

**Comment by Xrayez** [comment]
> @Xrayez Sorry, I don't pay attention to Goost, I have no idea what is going on in there.

No worries, I can explain this in more detail, if you want.

> The linker builds a dependency tree between symbols and anything not connected to the entry point is removed.

Does it happen only when compiling with `use_lto=yes` or regular debug builds as well? In Goost, it's not only about optimizing for size, it's also about skipping compilation of disabled classes altogether, so it also saves build time.

Just like in this PR, I also have to define `GOOST_ClassName` macros of those classes which are enabled, but not necessarily because of defining which classes got disabled this way, but because there are cases where disabling a single class leads to build error due to dependency, so I have to compile out stuff this way conditionally.

Much like current modules can be disabled and cause build errors. Similarly, Godot defines `MODULE_X_ENABLED` macros to conditionally compile out those extra features that depend on modules functionality, which do lead to various build errors according to #39196.

Modules also have a mini-dependency system implemented in Godot already, for those kind of dependencies which cannot be resolved using preprocessor instructions, see https://github.com/godotengine/godot/blob/c759b7b235513f6e67fa17bafb1f205b38eff060/methods.py#L305-L326

So I'd expect a similar mechanism to be present when disabling individual classes, so that people know that those cannot be disabled when they try to build Godot immediately. Unless you're OK with people trying to disable classes which are not supposed to be disabled, like `ClassDB` itself. You may also expect bug reports flooding the issue tracker because of this.

TL;DR: How would you solve the case when a user disables a base class? This will certainly lead to either compile or runtime error.

**Comment by reduz** [comment]
@Xrayez 
> Does it happen only when compiling with use_lto=yes or regular debug builds as well? In Goost, it's not only about optimizing for size, it's also about skipping compilation of disabled classes altogether, so it also saves build time.

It happens always, and you don't need to skip class compilation. It's in fact better if this does not happen as classes may still be used internally by the engine, even if not by the binder and you won't have a problem of something not compiling.

**Comment by Xrayez** [comment]
Since Godot is not a library and everything is basically treated to be internal on C++ level, this probably makes sense. Yet if Godot aims to make and organize existing code in a more modular way, this is just something to consider. This is probably not a big issue considering that Godot is small enough to not worry about this.

For instance, if you disable all physics-related classes, you may as well strip physics server code to achieve even smaller binary sizes, and whatnot. But for this, I've introduced the concept of [*Components* in Goost](https://goost.readthedocs.io/en/latest/development/components.html), which represent fairly distinct pieces of functionality.

Unlike modules, components are based on type of features in general, not content. Modules can also be part of components, for instance. Individual classes are also assigned to specific components in Goost. I do realize that modules can be used for this just fine, yet many existing features in Godot which are not implemented via modules (like in `scene/`) cannot be easily stripped, including those backends that may be still initialized at entry point, even if no front-end classes are enabled.

If we take animation system, in theory it could also be disabled entirely from the engine. But that depends on whether you treat animation stuff as core.

This is going slightly off-topic, but just wanted to brainstorm a bit, since we talk about customizations here (do we have a proposal about this, by the way)?

**Comment by reduz** [comment]
@Xrayez As I mentioned, I am not going to do much speculation for the time being, all I want to do is implement a framework for disabling classes. How they are going to be efficiently disabled will likely be part of a separate PR/proposal/etc.

**Comment by neikeq** [comment]
Sorry, after a second look this is actually fine, but I think the option should be `enable_classes`. I think it's simpler for the editor to notify of what it uses and leave knowledge of what it doesn't to the linker.

> Eventually, a proper UI will be implemented to create a build config file to do this at large scale, as well as detect what is used in the project. Automatic detection of classes used to generate the right build config is also a possibility.

I think we should definitely aim at automatic detection in scene files, resource files, scripts, etc. This would benefit code trimming in the C# bindings as well (2 MB total when not trimmed, not too much but worth reducing).

**Comment by neikeq** [comment]
I don't think there's anything that prevents use from doing automatic detection of most things. We can read scene/resource files, and at least for GDScript it should be possible to get the classes being used from the parser. Only exception is "reflection", as in code that deals with ClassDB via strings.

**Comment by reduz** [comment]
@neikeq I prefer to pass the ones that are disabled, so if we have a new class is added afterwards and for some reason the build config was not updated, it wont stop working for some random reason.

**Comment by lyuma** [comment]
A bit of a user question... As someone who is creating addons to publish in the AssetLib, how can I make my addons tolerate missing classes while still providing some functionality.

Currently, referencing a type that doesn't exist produces a compile time error, even if the class use is surrounded in a `type_exists()` call:
image

I could instead use `ClassDB.instance("SomeType")` to construct instances, and avoid any type annotations to specific node types that I fear could be compiled out... but it's a bit annoying to do this everywhere.

Put another way, I'm worried about fragmentation in Godot's asset ecosystem when everyone may have a different set of classes disabled.

**Comment by reduz** [comment]
@lyuma what will happen here in theory is the opposite, the editor will scan your add-on, and generate a build configuration that keeps thr classes it uses.

**Comment by aaronfranke** [comment]
@lyuma To clarify what @reduz said: The Godot editor in the future will be able to scan entire Godot projects for the purpose of  generating a config for optimized export templates, but the editor itself will still have everything. If users install your addon, the scanning tool will scan the entire project, including your addon, so it won't exclude anything your addon needs.

**Comment by lyuma** [comment]
Thanks for the explanations, @aaronfranke @reduz I understand. It will be completely transparent by only removing unused features from the engine.
Also, as it only affects export templates, tool addons won't be affected at all.

File: core/typedefs.h [hunk 1]
```diff
@@ -281,4 +281,11 @@ struct BuildIndexSequence<0, Is...> : IndexSequence<Is...> {};
 #define DEBUG_METHODS_ENABLED
 #endif
 
+//macro GD_IS_DEFINED() allows to check if a macro is defined. It needs to be defined to anything (say 1) to work.
```
Comment (akien-mga) [inline core/typedefs.h:284]:
```suggestion
// Macro GD_IS_DEFINED() allows to check if a macro is defined. It needs to be defined to anything (say 1) to work.
```

File: core/typedefs.h [hunk 2]
```diff
@@ -281,4 +281,11 @@ struct BuildIndexSequence<0, Is...> : IndexSequence<Is...> {};
...
+#define __GDARG_PLACEHOLDER_1 0,
+#define __gd_take_second_arg(__ignored, val, ...) val
+#define GD_IS_DEFINED(x) ___gd_is_defined(x)
+#define ___gd_is_defined(val) ____gd_is_defined(__GDARG_PLACEHOLDER_##val)
+#define ____gd_is_defined(arg1_or_junk) __gd_take_second_arg(arg1_or_junk 1, 0)
```
Comment (akien-mga) [inline core/typedefs.h:289]: Nitpick but this might read better:
```suggestion
#define __gd_take_second_arg(__ignored, val, ...) val
#define ____gd_is_defined(arg1_or_junk) __gd_take_second_arg(arg1_or_junk 1, 0)
#define __GDARG_PLACEHOLDER_1 0,
#define ___gd_is_defined(val) ____gd_is_defined(__GDARG_PLACEHOLDER_##val)
#define GD_IS_DEFINED(x) ___gd_is_defined(x)
```
so you use them in the right order.

**Review by akien-mga — APPROVED** [review]
One typo that would break with `disable_classes="Node2D, Node3D"`, otherwise seems good.

**Comment by akien-mga** [comment]
Thanks!

**Comment by dreamsComeTrue** [comment]
Would be a tremendous work if someone could calculate, how much space does each Class/Node take :)

I mean - does it save that much in the end? I would say, modules are the one which probably takes most of the space (for example: get rid of FBX import if not used), but I am not so sure if the space-saving-rescue comes from the Nodes themselves...

**Comment by Xrayez** [comment]
> Would be a tremendous work if someone could calculate, how much space does each Class/Node take :)

It's probably not a lot, I think you'll only benefit from this if you have a relatively small project which doesn't use most of the built-in classes. This is one of the reasons why I was suggesting that more work could be done into organizing Godot's core features into components, because many structures are still hardcoded like physics server and cannot be completely compiled out even if you don't use physics in your game. Just like we have ability to pass `disable_advanced_gui=yes` etc. when compiling the engine, but lets say it would be `physics_enabled=no`.
